File: examples/trenchOxidation/trenchOxidation.py
#!/usr/bin/env python3
"""
Trench Structure Oxidation Example (ViennaPS)
=============================================
Simulates thermal oxidation of a rectangular Si trench etched into a flat
Si substrate. Oxide grows on all exposed Si surfaces: trench sidewalls,
trench floor, and the surrounding flat substrate.

Geometry (2-D cross-section, extruded symmetrically in Z for 3-D):
    · Flat Si substrate at y = 0
    · Trench centered at x = 0: width trenchWidth, depth trenchDepth (into Si)

Usage:
    python trenchOxidation.py [config.txt]

All lengths are in micrometers, time in hours, pressure in atm.
"""

import sys
import time
import viennaps as vps

# ── Default parameters (match trenchOxidation/config.txt) ───────────────────
cfg = {
    "dimensions":    2,
    "numThreads":    16,
    "gridDelta":     0.05,
    "xExtent":       0.6,
    "zExtent":       0.0,   # 3D only: half-depth in Z; defaults to xExtent if 0
    "trenchWidth":   0.3,
    "trenchDepth":   0.5,
    "oxideThickness": 0.0,
    "oxidationTime": 0.2,
    "temperature": 1000.0,
    "pressure":      1.0,
    "oxidant":            "wet",
    "orientation":        "100",
    "outputPrefix":       "ps_trench_oxidation",
    "maxGridPoints":      0,   # 0 = unlimited; set >0 to cap the Cartesian solve grid
    "useGpu":             "auto",   # auto | gpu | cpu
    "gpuPreconditioner":  "jacobi", # jacobi | ilu0
}


def _parse_config(path: str) -> None:
    try:
        with open(path) as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                eq = line.find("=")
                if eq < 0:
                    continue
                key = line[:eq].strip()
                val = line[eq + 1:].split("#")[0].strip()
                if key in cfg:
                    t = type(cfg[key])
                    cfg[key] = t(val)
    except FileNotFoundError:
        pass


def _parse_oxidant(s: str):
    s = s.lower()
    if s in ("wet", "h2o"):
        return vps.OxidantType.Wet
    if s in ("dry", "o2"):
        return vps.OxidantType.Dry
    raise ValueError(f"Unknown oxidant '{s}'. Use wet/H2O or dry/O2.")


def _parse_orientation(s: str):
    s = s.lower()
    if s in ("100", "<100>", "si100"):
        return vps.SiliconOrientation.Si100
    if s in ("110", "<110>", "si110"):
        return vps.SiliconOrientation.Si110
    if s in ("111", "<111>", "si111"):
        return vps.SiliconOrientation.Si111
    if s in ("poly", "polysi"):
        return vps.SiliconOrientation.PolySi
    raise ValueError(f"Unknown orientation '{s}'. Use 100, 110, 111, or poly.")


# ── Config file ───────────────────────────────────────────────────────────────
config_file = sys.argv[1] if len(sys.argv) > 1 else "config.txt"
_parse_config(config_file)

vps.setDimension(cfg["dimensions"])
vps.setNumThreads(cfg["numThreads"])
vps.Logger.setLogLevel(vps.LogLevel.ERROR)

grid_delta      = cfg["gridDelta"]
x_extent        = cfg["xExtent"]
z_extent        = cfg["zExtent"] if cfg["zExtent"] > 0.0 else cfg["xExtent"]
trench_width    = cfg["trenchWidth"]
trench_depth    = cfg["trenchDepth"]
oxide_thickness = cfg["oxideThickness"]
oxidation_time  = cfg["oxidationTime"]
temperature     = cfg["temperature"]
pressure        = cfg["pressure"]
oxidant         = _parse_oxidant(cfg["oxidant"])
orientation     = _parse_orientation(cfg["orientation"])
output_prefix   = cfg["outputPrefix"]

# ── Build Si trench geometry ──────────────────────────────────────────────────
# MakeTrench's second constructor sets up the domain (REFLECTIVE X, INFINITE Y)
# and creates the trench: flat substrate at y=0 with a rectangular slot of
# width trenchWidth and depth trenchDepth centered at x=0.
domain = vps.Domain()
y_extent = 2.0 * z_extent if cfg["dimensions"] == 3 else 0.0
vps.MakeTrench(domain, gridDelta=grid_delta, xExtent=2.0 * x_extent, yExtent=y_extent,
               trenchWidth=trench_width, trenchDepth=trench_depth).apply()

# ── Oxide seed ────────────────────────────────────────────────────────────────
import viennals as vls

seed_thickness = max(oxide_thickness, grid_delta)
ambient_ls = vls.Domain(domain.getLevelSets()[-1])
vls.GeometricAdvect(ambient_ls, vls.SphereDistribution(seed_thickness)).apply()
domain.insertNextLevelSetAsMaterial(ambient_ls, vps.Material.SiO2, False)

# ── Oxidation model ───────────────────────────────────────────────────────────
model = vps.Oxidation()
model.setTemperature(temperature)
model.setTime(oxidation_time)
model.setOxidant(oxidant)
model.setPressure(pressure)
model.setOrientation(orientation)
model.setInitialOxideThickness(seed_thickness)

use_gpu = cfg["useGpu"].lower()
if use_gpu == "gpu":
    model.setGpuMode(vps.GpuMode.Gpu)
elif use_gpu == "cpu":
    model.setGpuMode(vps.GpuMode.Cpu)
if cfg["gpuPreconditioner"].lower() == "ilu0":
    model.setGpuPreconditioner(vps.GpuPreconditioner.ILU0)

if cfg["maxGridPoints"] > 0:
    model.setMaxGridPoints(int(cfg["maxGridPoints"]))

model.saveSurfaceMesh(domain, output_prefix + "_initial.vtp")
model.saveVolumeMesh(domain, output_prefix + "_initial")

t0 = time.perf_counter()
vps.Process(domain, model, 0.0).apply()
elapsed_sim = time.perf_counter() - t0

model.saveSurfaceMesh(domain, output_prefix + "_after.vtp")
model.saveVolumeMesh(domain, output_prefix + "_after")

print(f"Simulation time: {elapsed_sim:.2f} s")
print(
    f"Planar Deal-Grove estimate for {oxidation_time} hr at {temperature} °C: "
    f"{model.estimatePlanarOxideThickness(seed_thickness):.4f} µm oxide."
)
print(
    f"Wrote {output_prefix}_initial.vtp, {output_prefix}_after.vtp, "
    f"and {output_prefix}_after_volume.vtu"
)

File: examples/trenchOxidation/trenchOxidation.cpp
#include <models/psOxidation.hpp>
#include <process/psProcess.hpp>
#include <psDomain.hpp>
#include <psUtil.hpp>
#include <geometries/psMakeTrench.hpp>

#include <lsGeometricAdvect.hpp>

#include <algorithm>
#include <array>
#include <cctype>
#include <chrono>
#include <iostream>
#include <omp.h>
#include <stdexcept>
#include <string>

namespace ps = viennaps;
namespace ls = viennals;

using NumericType = double;

// ---------------------------------------------------------------------------
// Utility helpers
// ---------------------------------------------------------------------------

std::string lower(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(),
                 [](unsigned char c) { return std::tolower(c); });
  return value;
}

std::string getString(const ps::util::Parameters &params, const char *key,
                      const std::string &fallback) {
  const auto it = params.m.find(key);
  return it == params.m.end() ? fallback : it->second;
}

ps::OxidantType parseOxidant(const std::string &value) {
  const auto n = lower(value);
  if (n == "wet" || n == "h2o") return ps::OxidantType::Wet;
  if (n == "dry" || n == "o2")  return ps::OxidantType::Dry;
  throw std::invalid_argument("Unknown oxidant '" + value +
                              "'. Use wet/H2O or dry/O2.");
}

ps::SiliconOrientation parseOrientation(const std::string &value) {
  const auto n = lower(value);
  if (n == "100" || n == "<100>" || n == "si100") return ps::SiliconOrientation::Si100;
  if (n == "110" || n == "<110>" || n == "si110") return ps::SiliconOrientation::Si110;
  if (n == "111" || n == "<111>" || n == "si111") return ps::SiliconOrientation::Si111;
  if (n == "poly" || n == "polysi" || n == "poly-silicon")
    return ps::SiliconOrientation::PolySi;
  throw std::invalid_argument("Unknown orientation '" + value +
                              "'. Use 100, 110, 111, or poly.");
}

// ---------------------------------------------------------------------------
// Simulation driver
//
// Coordinate convention:
//   2D: X = lateral (REFLECTIVE), Y = growth (INFINITE)
//   3D: X = lateral (REFLECTIVE), Y = trench extrusion (REFLECTIVE), Z = growth (INFINITE)
//       yMin/yMax refer to the growth axis (Z in 3D); zExtent is the Y extrusion range.
// ---------------------------------------------------------------------------

template <int D>
void run(const ps::util::Parameters &params) {
  omp_set_num_threads(params.get<int>("numThreads"));
  ps::Logger::setLogLevel(ps::LogLevel::ERROR);

  const NumericType gridDelta     = params.get("gridDelta");
  const NumericType xExtent       = params.get("xExtent");
  const NumericType yMin          = params.get("yMin");
  const NumericType yMax          = params.get("yMax");
  const NumericType trenchWidth   = params.get("trenchWidth");
  const NumericType trenchDepth   = params.get("trenchDepth");
  const auto oxIt = params.m.find("oxideThickness");
  const NumericType oxideThickness =
      (oxIt != params.m.end()) ? params.get("oxideThickness") : NumericType(0);
  const NumericType oxidationTime = params.get("oxidationTime");
  const NumericType temperature   = params.get("temperature");
  const NumericType pressure      = params.get("pressure");

  const NumericType zExtent = [&]() -> NumericType {
    if constexpr (D != 3) return NumericType(0);
    const auto it = params.m.find("zExtent");
    return (it == params.m.end()) ? xExtent : params.get("zExtent");
  }();

  const auto oxidant      = parseOxidant(getString(params, "oxidant", "wet"));
  const auto orientation  = parseOrientation(getString(params, "orientation", "100"));
  const auto outputPrefix = getString(params, "outputPrefix", "ps_trench_oxidation");

  using BoundaryType = typename ls::Domain<NumericType, D>::BoundaryType;
  double bounds[2 * D];
  BoundaryType boundaryCons[D];

  bounds[0] = -xExtent; bounds[1] = xExtent;
  if constexpr (D == 2) {
    bounds[2] = yMin; bounds[3] = yMax;
    boundaryCons[0] = BoundaryType::REFLECTIVE_BOUNDARY;
    boundaryCons[1] = BoundaryType::INFINITE_BOUNDARY;
  } else {
    // Y = trench extrusion (REFLECTIVE), Z = growth (INFINITE)
    bounds[2] = -zExtent; bounds[3] = zExtent;
    bounds[4] = yMin;     bounds[5] = yMax;
    boundaryCons[0] = BoundaryType::REFLECTIVE_BOUNDARY;
    boundaryCons[1] = BoundaryType::REFLECTIVE_BOUNDARY;
    boundaryCons[2] = BoundaryType::INFINITE_BOUNDARY;
  }

  auto domain = ps::Domain<NumericType, D>::New(bounds, boundaryCons, gridDelta);
  ps::MakeTrench<NumericType, D>(domain, trenchWidth, trenchDepth).apply();

  // Clamp the oxide seed to at least gridDelta so the Cartesian solve always
  // has resolvable nodes between the Si and SiO2 level sets.
  const NumericType seedThickness = std::max(oxideThickness, gridDelta);
  {
    auto ambientInterface = ls::Domain<NumericType, D>::New(domain->getLevelSets().back());
    auto initialOxide =
        ps::SmartPointer<ls::SphereDistribution<viennahrle::CoordType, D>>::New(
            seedThickness);
    ls::GeometricAdvect<NumericType, D>(ambientInterface, initialOxide).apply();
    domain->insertNextLevelSetAsMaterial(ambientInterface, ps::Material::SiO2,
                                         false);
  }

  auto model = ps::SmartPointer<ps::Oxidation<NumericType, D>>::New();
  model->setTemperature(temperature);
  model->setTime(oxidationTime);
  model->setOxidant(oxidant);
  model->setPressure(pressure);
  model->setOrientation(orientation);
  model->setInitialOxideThickness(seedThickness);

  {
    const auto useGpu = lower(getString(params, "useGpu", "auto"));
    if      (useGpu == "gpu") model->setGpuMode(ps::GpuMode::Gpu);
    else if (useGpu == "cpu") model->setGpuMode(ps::GpuMode::Cpu);
    const auto prec = lower(getString(params, "gpuPreconditioner", "jacobi"));
    if (prec == "ilu0") model->setGpuPreconditioner(ps::GpuPreconditioner::ILU0);
  }

  {
    const auto it = params.m.find("maxGridPoints");
    if (it != params.m.end())
      model->setMaxGridPoints(
          static_cast<std::size_t>(std::stoull(it->second)));
  }

  model->saveSurfaceMesh(domain, outputPrefix + "_stack_initial.vtp");
  model->saveVolumeMesh(domain, outputPrefix + "_stack_initial");

  const auto t0 = std::chrono::steady_clock::now();
  ps::Process<NumericType, D>(domain, model, NumericType(0)).apply();
  const double elapsedSim =
      std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();

  model->saveSurfaceMesh(domain, outputPrefix + "_stack_after.vtp");
  model->saveVolumeMesh(domain, outputPrefix + "_stack_after");

  std::cout << "Simulation time: " << elapsedSim << " s\n";
  std::cout << "Planar Deal-Grove estimate for " << oxidationTime
            << " hr oxidation at " << temperature << " C: "
            << model->estimatePlanarOxideThickness(seedThickness)
            << " um oxide thickness." << std::endl;

  std::cout << "Wrote " << outputPrefix << "_stack_initial.vtp and "
            << outputPrefix << "_stack_after.vtp and "
            << outputPrefix << "_stack_after_volume.vtu" << std::endl;
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

int main(int argc, char **argv) {
  ps::util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    params.readConfigFile("config.txt");
    if (params.m.empty()) {
      std::cout << "No configuration file provided!" << std::endl;
      std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
      return 1;
    }
  }

  const int dimensions = std::stoi(getString(params, "dimensions", "2"));
  if (dimensions == 3)
    run<3>(params);
  else
    run<2>(params);

  return 0;
}
